src/coreclr/src/jit/morph.cpp

```diff
@@ -7298,306 +7283,724 @@ GenTree* Compiler::fgMorphPotentialTailCall(GenTreeCall* call)
         }
 
-        vtbl = gtNewOperNode(GT_IND, TYP_I_IMPL, add);
+        if (pRuntimeLookup->offsets[i] != 0)
+        {
+            result = gtNewOperNode(GT_ADD, TYP_I_IMPL, result, gtNewIconNode(pRuntimeLookup->offsets[i], TYP_I_IMPL));
+        }
+    }
 
-        if (isRelative)
+    // No null test required
+    if (!pRuntimeLookup->testForNull)
+    {
+        if (pRuntimeLookup->indirections > 0)
         {
-            vtbl = gtNewOperNode(GT_ADD, TYP_I_IMPL, vtbl, indOffTree);
+            result = gtNewOperNode(GT_IND, TYP_I_IMPL, result);
+            result->gtFlags |= GTF_IND_NONFAULTING;
+
+            if (pRuntimeLookup->testForFixup)
+            {
+                unsigned slotLclNum = lvaGrabTemp(true DEBUGARG("getRuntimeLookupTree test"));
+                stmts.Push(gtNewTempAssign(slotLclNum, result));
+
+                GenTree* slot = gtNewLclvNode(slotLclNum, TYP_I_IMPL);
+                // downcast the pointer to a TYP_INT on 64-bit targets
+                slot = impImplicitIorI4Cast(slot, TYP_INT);
+                // Use a GT_AND to check for the lowest bit and indirect if it is set
+                GenTree* test  = gtNewOperNode(GT_AND, TYP_INT, slot, gtNewIconNode(1));
+                GenTree* relop = gtNewOperNode(GT_EQ, TYP_INT, test, gtNewIconNode(0));
+
+                // slot = GT_IND(slot - 1)
+                slot           = gtNewLclvNode(slotLclNum, TYP_I_IMPL);
+                GenTree* add   = gtNewOperNode(GT_ADD, TYP_I_IMPL, slot, gtNewIconNode(-1, TYP_I_IMPL));
+                GenTree* indir = gtNewOperNode(GT_IND, TYP_I_IMPL, add);
+                indir->gtFlags |= GTF_IND_NONFAULTING;
+                indir->gtFlags |= GTF_IND_INVARIANT;
+
+                slot           = gtNewLclvNode(slotLclNum, TYP_I_IMPL);
+                GenTree* asg   = gtNewAssignNode(slot, indir);
+                GenTree* colon = new (this, GT_COLON) GenTreeColon(TYP_VOID, gtNewNothingNode(), asg);
+                GenTree* qmark = gtNewQmarkNode(TYP_VOID, relop, colon);
+                stmts.Push(qmark);
+
+                result = gtNewLclvNode(slotLclNum, TYP_I_IMPL);
+            }
         }
+    }
+    else
+    {
+        assert(pRuntimeLookup->indirections != 0);
+
+        // Extract the handle
+        GenTree* handle = gtNewOperNode(GT_IND, TYP_I_IMPL, result);
+        handle->gtFlags |= GTF_IND_NONFAULTING;
+
+        GenTree* handleCopy = cloneTree(&handle DEBUGARG("getRuntimeLookupTree handle"));
+
+        // Call to helper
+        GenTree* argNode =
+            gtNewIconEmbHndNode(pRuntimeLookup->signature, nullptr, GTF_ICON_TOKEN_HDL, compileTimeHandle);
+
+        GenTreeCall::Use* helperArgs =
+            gtNewCallArgs(getRuntimeContextTree(pLookup->lookupKind.runtimeLookupKind), argNode);
+        GenTree* helperCall = gtNewHelperCallNode(pRuntimeLookup->helper, TYP_I_IMPL, helperArgs);
+
+        result = helperCall;
+        //// Check for null and possibly call helper
```

Comment:
I assume that this is a remaining "TODO"?

Reply:
Right, the problem I ran into is that we cannot create QMARK nodes at this point in morph. My plan is to try to use `GT_RUNTIMELOOKUP` to represent this entire thing and then expand this later in the JIT.

Reply:
@jakobbotsch Do you have a test case that will hit this code path?